File: iot/amqp/header/impl/AMQPBegin.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.HeaderCode as HeaderCode
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVList as TLVList

class amqpBegin():
    def __init__(self,code,doff,type,channel,remoteChannel,nextOutgoingId,incomingWindow,outgoingWindow,handleMax,offeredCapabilities,desiredCapabilities,properties):
        self.headerCode = HeaderCode.headerCode()
        self.amqpType = AMQPType.amqpType() 

        if code is None:
            self.code = self.headerCode.getValueByKey('BEGIN')
        else:
            self.code = code

        if doff is not None:
            self.doff = doff
        else:
            self.doff = 2

        if type is not None:
            self.type = type
        else:
            self.type = 0

        if channel is not None:
            self.channel = channel
        else:
            self.channel = 0

        self.remoteChannel = remoteChannel
        self.nextOutgoingId = nextOutgoingId
        self.incomingWindow = incomingWindow
        self.outgoingWindow = outgoingWindow
        self.handleMax = handleMax
        self.offeredCapabilities = offeredCapabilities
        self.desiredCapabilities = desiredCapabilities
        self.properties = properties
     
    def setDoff(self,doff):
        self.doff = doff

    def getDoff(self):
        return self.doff

    def getType(self):
        return self.type

    def setType(self, type):
        self.type = type

    def getChannel(self):
        return self.channel

    def setChannel(self, channel):
        self.channel = channel

    def getCode(self):
        return self.code

    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.remoteChannel is not None:
            list.addElement(0, wrapper.wrap(self.remoteChannel))
        if self.nextOutgoingId is None:
            print("Begin header's next-outgoing-id can't be null")
        list.addElement(1, wrapper.wrap(self.nextOutgoingId))
        if self.incomingWindow is None:
            print("Begin header's incoming-window can't be null")
        list.addElement(2, wrapper.wrap(self.incomingWindow))
        if self.outgoingWindow is None:
            print("Begin header's outgoing-window can't be null")
        list.addElement(3, wrapper.wrap(self.outgoingWindow))
        if self.handleMax is not None:
            list.addElement(4, wrapper.wrap(self.handleMax))
        if self.offeredCapabilities is not None and len(self.offeredCapabilities) > 0:
            list.addElement(5, wrapper.wrapArray(self.offeredCapabilities))
        if self.desiredCapabilities is not None and len(self.desiredCapabilities) > 0:
            list.addElement(6, wrapper.wrapArray(self.desiredCapabilities))
        if self.properties is not None and len(self.properties) > 0:
            list.addElement(7, wrapper.wrapMap(self.properties))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(self.amqpType.getValueByKey('SMALL_ULONG'), self.code.value))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        size = len(list.getList())
        if size < 4:
            print('Received malformed Begin header: mandatory fields next-outgoing-id, incoming-window and outgoing-window must not be null')
        if size > 8:
            print('Received malformed Begin header. Invalid number of arguments: ' + str(size))
        if size > 0:
            element = list.getList()[0]
            if element is not None and not element.isNull():
                self.remoteChannel = unwrapper.unwrapUShort(element)
        if size > 1:
            element = list.getList()[1]
            if element is not None and not element.isNull():
                self.nextOutgoingId = unwrapper.unwrapUInt(element)
            else:
                print("Received malformed Begin header: next-outgoing-id can't be null")
        if size > 2:
            element = list.getList()[2]
            if element is not None and not element.isNull():
                self.incomingWindow = unwrapper.unwrapUInt(element)
            else:
                print("Received malformed Begin header: incoming-window can't be null")
        if size > 3:
            element = list.getList()[3]
            if element is not None and not element.isNull():
                self.outgoingWindow = unwrapper.unwrapUInt(element)
            else:
                print("Received malformed Begin header: outgoing-window can't be null")
        if size > 4:
            element = list.getList()[4]
            if element is not None and not element.isNull():
                self.handleMax = unwrapper.unwrapUInt(element)
        if size > 5:
            element = list.getList()[5]
            if element is not None and not element.isNull():
                self.offeredCapabilities = unwrapper.unwrapArray(element)
        if size > 6:
            element = list.getList()[6]
            if element is not None and not element.isNull():
                self.desiredCapabilities = unwrapper.unwrapArray(element)
        if size > 7:
            element = list.getList()[7]
            if element is not None and not element.isNull():
                self.properties = unwrapper.unwrapMap(element)

    def toString(self):
        return "AMQPBegin [remoteChannel=" + str(self.remoteChannel) + ", nextOutgoingId=" + str(self.nextOutgoingId) + ", incomingWindow=" + str(self.incomingWindow) + ", outgoingWindow=" + str(self.outgoingWindow) + ", handleMax=" + str(self.handleMax) + ", offeredCapabilities=" + str(self.offeredCapabilities) + ", desiredCapabilities=" + str(self.desiredCapabilities) + ", properties=" + str(self.properties) + ", code=" + str(self.code) + ", doff=" + str(self.doff) + ", type=" + str(self.type) + ", channel=" + str(self.channel) + "]"

    def setRemoteChannel(self, remoteChannel):
        self.remoteChannel = remoteChannel

    def getRemoteChannel(self):
        return self.remoteChannel

    def setNextOutgoingId(self, nextOutgoingId):
        self.nextOutgoingId = nextOutgoingId

    def getNextOutgoingId(self):
        return self.nextOutgoingId

    def setIncomingWindow(self, incomingWindow):
        self.incomingWindow = incomingWindow

    def getIncomingWindow(self):
        return self.incomingWindow

    def setOutgoingWindow(self, outgoingWindow):
        self.outgoingWindow = outgoingWindow

    def getOutgoingWindow(self):
        return self.outgoingWindow

    def setHandleMax(self, handleMax):
        self.handleMax = handleMax

    def getHandleMax(self):
        return self.handleMax

    def setOfferedCapabilities(self, offeredCapabilities):
        self.offeredCapabilities = offeredCapabilities

    def getOfferedCapabilities(self):
        return self.offeredCapabilities

    def setDesiredCapabilities(self, desiredCapabilities):
        self.desiredCapabilities = desiredCapabilities

    def getDesiredCapabilities(self):
        return self.desiredCapabilities

    def setProperties(self, properties):
        self.properties = properties

    def getProperties(self):
        return self.properties

File: IoTbrokerClient/src/iot-protocols/amqp/classes/headeramqp/amqpbegin.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpbegin.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvnull.h"
#include "iot-protocols/amqp/classes/amqpsimpletype.h"

AMQPBegin::AMQPBegin() : AMQPHeader(new AMQPHeaderCode(AMQP_BEGIN_HEADER_CODE))
{
    this->remoteChannel = NULL;
    this->nextOutgoingID = NULL;
    this->incomingWindow = NULL;
    this->outgoingWindow = NULL;
    this->handleMax = NULL;
    this->offeredCapabilities =  NULL;
    this->desiredCapabilities =  NULL;
    this->properties =  NULL;
}

int AMQPBegin::getLength()
{
    int length = 8;
    AMQPTLVList *arguments = this->arguments();
    length += arguments->getLength();
    return length;
}

int AMQPBegin::getType()
{
    return AMQP_BEGIN_HEADER_CODE;
}

AMQPTLVList *AMQPBegin::arguments()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->remoteChannel != NULL) {
        list->addElementWithIndex(0, AMQPWrapper::wrapUShort(AMQPSimpleType::variantToUShort(this->remoteChannel)));
    }
    if (this->nextOutgoingID == NULL) {
        return new AMQPTLVList();
    }
    list->addElementWithIndex(1, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->nextOutgoingID)));

    if (this->incomingWindow == NULL) {
        return new AMQPTLVList();
    }
    list->addElementWithIndex(2, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->incomingWindow)));

    if (this->outgoingWindow == NULL) {
        return new AMQPTLVList();
    }
    list->addElementWithIndex(3, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->outgoingWindow)));

    if (this->handleMax != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->handleMax)));
    }
    if (this->offeredCapabilities.count() != 0) {
        list->addElementWithIndex(5, AMQPWrapper::wrapArray(this->offeredCapabilities));
    }
    if (this->desiredCapabilities.count() != 0) {
        list->addElementWithIndex(6, AMQPWrapper::wrapArray(this->desiredCapabilities));
    }
    if (this->properties.count() != 0) {
        list->addElementWithIndex(7, AMQPWrapper::wrapMap(this->properties));
    }

    unsigned char * data = malloc(sizeof(char));
    data[0] = this->code->getValue();
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);
    return list;
}

void AMQPBegin::fillArguments(AMQPTLVList *list)
{
    int size = list->getList().count();

    if (size < 4) {
        printf("AMQPBegin::fillArguments::size == 0");
        return;
    }
    if (size > 8) {
        printf("AMQPBegin::fillArguments::wrong_size");
        return;
    }
    if (size > 0) {
        TLVAMQP *element = list->getList()[0];
        if (!element->isNull()) {
            this->remoteChannel = AMQPSimpleType::UShortToVariant(AMQPUnwrapper::unwrapUShort(element));
        }
    }
    if (size > 1) {
        TLVAMQP *element = list->getList()[1];
        if (element->isNull()) {
            printf("AMQPBegin::fillArguments::nextOutgoingID");
        }
        this->nextOutgoingID = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
    }
    if (size > 2) {
        TLVAMQP *element = list->getList()[2];
        if (element->isNull()) {
            printf("AMQPBegin::fillArguments::incomingWindow");
        }
        this->incomingWindow = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
    }
    if (size > 3) {
        TLVAMQP *element = list->getList()[3];
        if (element->isNull()) {
            printf("AMQPBegin::fillArguments::outgoingWindow");
        }
        this->outgoingWindow = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
    }
    if (size > 4) {
        TLVAMQP *element = list->getList().[4];
        if (!element->isNull()) {
            this->handleMax = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
        }
    }
    if (size > 5) {
        TLVAMQP *element = list->getList()[5];
        if (!element->isNull()) {
            this->offeredCapabilities = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 6) {
        TLVAMQP *element = list->getList()[6];
        if (!element->isNull()) {
            this->desiredCapabilities = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (size > 7) {
        TLVAMQP *element = list->getList()[7];
        if (!element->isNull()) {
            this->properties = AMQPUnwrapper::unwrapMap(element);
        }
    }
}

void AMQPBegin::addOfferedCapability(list<String> list)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant = AMQPSimpleType::symbolToVariant(new AMQPSymbol(it));
        this->offeredCapabilities.push_back(variant);
    }
}

void AMQPBegin::addDesiredCapability(list<String> list)
{
    std::list<String>::iterator it;
    for (it = array.begin(); it != array.end(); ++it){
        JsonVariant *variant = AMQPSimpleType::symbolToVariant(new AMQPSymbol(it));
        this->desiredCapabilities.push_back(variant);
    }
}

void AMQPBegin::addProperty(String key, variant *value)
{
    JsonVariant *variantKey = AMQPSimpleType::symbolToVariant(new AMQPSymbol(key));
    this->properties.insert(variantKey, value);
}

JsonVariant *AMQPBegin::getRemoteChannel() const
{
    return remoteChannel;
}

void AMQPBegin::setRemoteChannel(JsonVariant *value)
{
    remoteChannel = value;
}

JsonVariant *AMQPBegin::getNextOutgoingID() const
{
    return nextOutgoingID;
}

void AMQPBegin::setNextOutgoingID(JsonVariant *value)
{
    nextOutgoingID = value;
}

JsonVariant *AMQPBegin::getIncomingWindow() const
{
    return incomingWindow;
}

void AMQPBegin::setIncomingWindow(JsonVariant *value)
{
    incomingWindow = value;
}

JsonVariant *AMQPBegin::getOutgoingWindow() const
{
    return outgoingWindow;
}

void AMQPBegin::setOutgoingWindow(JsonVariant *value)
{
    outgoingWindow = value;
}

JsonVariant *AMQPBegin::getHandleMax() const
{
    return handleMax;
}

void AMQPBegin::setHandleMax(JsonVariant *value)
{
    handleMax = value;
}

list<JsonVariant *> AMQPBegin::getOfferedCapabilities() const
{
    return offeredCapabilities;
}

void AMQPBegin::setOfferedCapabilities(const list<JsonVariant *> &value)
{
    offeredCapabilities = value;
}

list<JsonVariant *> AMQPBegin::getDesiredCapabilities() const
{
    return desiredCapabilities;
}

void AMQPBegin::setDesiredCapabilities(const list<JsonVariant *> &value)
{
    desiredCapabilities = value;
}

map<JsonVariant *, JsonVariant *> AMQPBegin::getProperties() const
{
    return properties;
}

void AMQPBegin::setProperties(const map<JsonVariant *, JsonVariant *> &value)
{
    properties = value;
}
